commons-impl/src/main/java/com/connexta/transformation/commons/api/impl/AbstractMetadataImpl.java

```diff
@@ -0,0 +1,357 @@
+  }
+
+  /**
+   * Gets the clock to use for retrieving wall and monotonic times.
+   *
+   * @return the clock to use for retrieving wall and monotonic times
+   */
+  public Clock getClock() {
+    return clock;
+  }
+
+  @Override
+  public int hashCode() {
+    return hashCode0();
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    return equals0(obj);
+  }
+
+  @Override
+  public String toString() {
+    return String.format(
+        "AbstractMetadataImpl[id=%s, transformId=%s, type=%s, requestInfo=%s, state=%s, startTime=%s, completionTime=%s, contentType=%s, contentLength=%d, failureReason=%s, failureMessage=%s]",
+        getId(),
+        transformId,
+        type,
+        requestInfo,
+        state,
+        startTime,
+        completionTime,
+        contentType,
+        contentLength,
+        failureReason,
+        failureMessage);
+  }
+
+  @Override
+  protected MetadataPojo writeTo(MetadataPojo pojo) throws PersistenceException {
+    if (hasUnknowns()) { // cannot serialize if it contains unknowns
+      throw new InvalidFieldException("unknown metadata");
+    }
+    super.writeTo(pojo);
+    setOrFailIfNullOrEmpty("transformId", this::getTransformId, pojo::setTransformId);
+    setOrFailIfNullOrEmpty("type", this::getMetadataType, pojo::setType);
+    convertAndSetOrFailIfNull(
+        "requestInfo", this::getRequestInfo, RequestInfoImpl::toPojo, pojo::setRequestInfo);
+    setOrFailIfNull("startTime", this::getStartTime, pojo::setStartTime);
+    convertAndSetOrFailIfNull("state", this::getState, State::name, pojo::setState);
+    convertAndSet(
+        "failureReason", this::getFailureReason0, ErrorCode::name, pojo::setFailureReason);
+    return pojo.setVersion(TransformationPojo.CURRENT_VERSION)
+        .setCompletionTime(completionTime)
+        .setFailureMessage(failureMessage)
+        .setContentType(contentType)
+        .setContentLength(Math.max(-1L, contentLength));
+  }
+
+  @Override
+  protected final void readFrom(MetadataPojo pojo) throws PersistenceException {
+    super.readFrom(pojo);
+    if (pojo.getVersion() < MetadataPojo.MINIMUM_VERSION) {
+      throw new UnsupportedVersionException(
+          "unsupported "
+              + AbstractMetadataImpl.PERSISTABLE_TYPE
+              + " version: "
+              + pojo.getVersion()
+              + " for object: "
+              + getId());
+    } // do support pojo.getVersion() > CURRENT_VERSION for forward compatibility
+    this.hasUnknowns = pojo instanceof UnknownPojo; // reset the unknown flag
+    readFromCurrentOrFutureVersion(pojo);
+  }
+
+  @VisibleForTesting
+  int hashCode0() {
+    return Objects.hash(
+        super.hashCode(),
+        type,
+        transformId,
+        state,
+        startTime,
+        completionTime,
+        failureReason,
+        failureMessage,
+        requestInfo,
+        contentType,
+        contentLength);
+  }
+
+  @VisibleForTesting
+  boolean equals0(Object obj) {
```

Comment:
This allows us to take advantage of Mockito's ability to create a mock for abstract classes without having to do it ourselves. Now, although when doing so you can tell Mockito to call the reals methods, it will never call `equals()` and `hashCode()` as it it will always provide an implementation of its own. Defining the actual logic in a private method (ok, package-private so it can be visible for testing), I can now test these methods easily.